### PR TITLE
Add selective signal method to support multicast

### PR DESCRIPTION
**Pull request by liebeym**
Some architecture needs to multicast  signal for subscriber / publisher model.
but XXX_signal connected emitter basically work on broadcast.

so adding selective signal method work on multicast with uniqueId list gathering from sender invocation data.

### CODE REVIEW

File: codegen_glibmm/templates/stub.cpp.templ [hunk 1]
```diff
@@ -201,6 +202,37 @@ void {{ class_name_with_namespace }}::{{ signal.name }}_emitter(
 }
 
 {% endfor %}
+
+{% for signal in interface.signals if signal is supported_by_sigc %}
+void {{ class_name_with_namespace }}::{{ signal.name }}_selectiveEmitter(
+  std::vector<Glib::ustring> &destination_bus_names,
```
Comment (mardy) [inline codegen_glibmm/templates/stub.cpp.templ]: indent it by 4 spaces, and add `const`

File: codegen_glibmm/templates/stub.h.templ [hunk 2]
```diff
@@ -57,6 +57,9 @@ protected:
 
     void {{ signal.name }}_emitter({{ signal.args|map(attribute='cpptype_out')|join(', ') }});
     sigc::signal<void, {{ signal.args|map(attribute='cpptype_out')|join(', ') }}> {{ signal.name }}_signal;
+
+    void {{ signal.name }}_selectiveEmitter(std::vector<Glib::ustring>&, {{ signal.args|map(attribute='cpptype_out')|join(', ') }});
```
Comment (mardy) [inline codegen_glibmm/templates/stub.h.templ]: What if we added an additional parameter to `emitter()`, instead (empty by default)?

File: codegen_glibmm/templates/stub.cpp.templ (hunk 1 above)
Comment (liebeym) [reply codegen_glibmm/templates/stub.cpp.templ]: I got it

File: codegen_glibmm/templates/stub.h.templ (hunk 2 above)
Comment (liebeym) [reply codegen_glibmm/templates/stub.h.templ]: Do you mean "emitter(const std::vector<Glib::ustring> &destination_bus_names = {}, ...)" default parameter?
So do you expect "XXX_signal.emit(DATA)" using broadcast and "XXX_signal.emit(ClientList, DATA)" using multicast ?

But, it will be binding signal to XXX_signal(mem_fun(this, &emitter), then it is required 
clientList parameter again like this "XXX_signal.emit({}, DATA)".
Because binded XXX_signal.emit() will don't know emitter function's default parameter. 
 
therefore I patched dedicated signal variable for multicast.

Comment (mardy) [reply codegen_glibmm/templates/stub.h.templ]: Yes I meant that, though I'd leave the list of destinations as second parameter, not as first one.

But now I see what the problem is. At the moment I cannot come up with any better solution, than what you already found, though I'm still not 100% conviced by it. Please let me wait for a few days, I wonder if we can come up with a better solution. @martin-ejdestig @erikboto @kursatkobya , any thoughts?

**Comment by liebeym** [comment]
@mardy 
Do you find the new method for emit?

Guys, Could you review this commit?
@martin-ejdestig @erikboto @kursatkobya

**Comment by liebeym** [comment]
@mardy 
Do I still wait this status? 
Are you find more good method?

File: codegen_glibmm/templates/stub.cpp.templ [hunk 3]
```diff
@@ -177,9 +178,10 @@ bool {{ class_name_with_namespace }}::on_interface_set_property(
 
 {% for signal in interface.signals if signal is supported_by_sigc %}
 void {{ class_name_with_namespace }}::{{ signal.name }}_emitter(
+    const std::vector<Glib::ustring> &destination_bus_names,
 {%- set comma = joiner() -%}
 {%- for arg in signal.args -%}
-{{ comma() }}{{ arg.cpptype_out }} {{ arg.name }}
+{{ comma() }}{{ arg.cpptype_out }} &{{ arg.name }}
```
Comment (mardy) [inline codegen_glibmm/templates/stub.cpp.templ]: The `&` should not be here. I suspect that the `arg.cpptype_out` is wrong, and that it should rather be `arg.cpptype_in`. But this is not really related with your changes, so, if you fix this, please do it in a separate commit.

File: codegen_glibmm/templates/stub.h.templ [hunk 4]
```diff
@@ -55,8 +55,27 @@ protected:
...
+        {{ signal.name }}_emitter(destination_bus_names, {{ arg.cpptype_to_dbus }}{{ arg.name }});
+        {% endfor %}
+    }
+    inline void {{ signal.name}}_broadcastEmitter(
+    {%- set comma = joiner() -%}
+    {%- for arg in signal.args -%}
+    {{ comma() }}{{ arg.cpptype_out }} {{ arg.name }}
+    {%- endfor -%}) {
+        {% for arg in signal.args %}
+        {{ signal.name }}_emitter({""}, {{ arg.cpptype_to_dbus }}{{ arg.name }});
```
Comment (mardy) [inline codegen_glibmm/templates/stub.h.templ]: I'd rather pass just `{}`, and let the emitter code take care of using an empty destination if an empty list of emitters is passed.

Comment (liebeym) [reply codegen_glibmm/templates/stub.h.templ]: I fixed it

**Comment by liebeym** [comment]
@mardy 
Is WRONG CI system? 
Test is all passed

File: codegen_glibmm/templates/stub.cpp.templ (hunk 3 above)
Comment (liebeym) [reply codegen_glibmm/templates/stub.cpp.templ]: Sure I will fix it to separate commit

File: codegen_glibmm/templates/stub.h.templ (hunk 4 above)
Comment (liebeym) [reply codegen_glibmm/templates/stub.h.templ]: @mardy 
I try recheck and consider about that, my patch was old ):

{} means to size 0 to vector container,
but, my patch is if "" is added to vector, broadcast emitter called because calling emitter(, null string,) is broadcasting.

so if vector container size is 0 by {}, emitter never called

File: codegen_glibmm/templates/stub.cpp.templ (hunk 3 above)
Comment (liebeym) [reply codegen_glibmm/templates/stub.cpp.templ]: I also patched other commit to this PR

File: codegen_glibmm/templates/stub.cpp.templ [hunk 5]
```diff
@@ -191,16 +193,19 @@ void {{ class_name_with_namespace }}::{{ signal.name }}_emitter(
...
-            "{{ signal.iface_name }}",
-            "{{ signal.name }}",
-            Glib::ustring(),
-            params);
+        for (const auto &bus_name: destination_bus_names) {
+            obj.connection->emit_signal(
+                    obj.object_path,
+                    "{{ signal.iface_name }}",
+                    "{{ signal.name }}",
+                    bus_name,
```
Comment (liebeym) [inline codegen_glibmm/templates/stub.cpp.templ]: @mardy 
I try recall and think about that again by my patch was old ):

{} means to size 0 to vector container,
so if "" is added to vector, bus_name = “” and emit_signal works on broadcast signal.
But if vector size is 0 by {}, emit_signal will be never called

**Comment by liebeym** [comment]
@mardy 
I was patch about your additional comment.
Could you review again?

File: codegen_glibmm/templates/stub.h.templ [hunk 6]
```diff
@@ -55,8 +55,27 @@ protected:
...
 
-    void {{ signal.name }}_emitter({{ signal.args|map(attribute='cpptype_out')|join(', ') }});
+    void {{ signal.name }}_emitter(const std::vector<Glib::ustring> &destination_bus_names, {{ signal.args|map(attribute='cpptype_out')|join(', ') }}&);
+    inline void {{ signal.name}}_multicastEmitter(const std::vector<Glib::ustring> &destination_bus_names,
+    {%- set comma = joiner() -%}
+    {%- for arg in signal.args -%}
+    {{ comma() }}{{ arg.cpptype_out }} {{ arg.name }}
+    {%- endfor -%}) {
+        {% for arg in signal.args %}
+        {{ signal.name }}_emitter(destination_bus_names, {{ arg.cpptype_to_dbus }}{{ arg.name }});
```
Comment (liebeym) [inline codegen_glibmm/templates/stub.h.templ]: I found a bug that multiple emitter function is generated when signal has a multiple argument.
 it will be change below like
  {{ signal.name }}_emitter(destination_bus_names,
                 {%- set comma = joiner() -%}
                 {%- for arg in signal.args -%}
                {{ comma() }}{{ arg.name }}
               {%- endfor -%});

File: codegen_glibmm/templates/stub.h.templ (hunk 4 above)
Comment (liebeym) [inline codegen_glibmm/templates/stub.h.templ]: It is also bug

**Comment by summeromance** [comment]
@liebeym How about declaring destination_bus_names as the last argument in the last order?
And we can decide whether to broadcast or multicast by checking the number of destination_bus_names in {signal.name} _emitter.

This way, there is no need to separate {signal.name} _broadcastEmitter and {signal.name} _multicastEmitter, and people who are already using generator can migrate without code modification.

**Comment by liebeym** [comment]
> @liebeym How about declaring destination_bus_names as the default argument in the last order?
> And we can decide whether to broadcast or multicast by checking the number of destination_bus_names in {signal.name} _emitter.
> 
> This way, there is no need to separate {signal.name} _broadcastEmitter and {signal.name} _multicastEmitter, and people who are already using generator can migrate without code modification.

**Comment by liebeym** [comment]
@s

> @liebeym How about declaring destination_bus_names as the default argument in the last order?
> And we can decide whether to broadcast or multicast by checking the number of destination_bus_names in {signal.name} _emitter.
> 
> This way, there is no need to separate {signal.name} _broadcastEmitter and {signal.name} _multicastEmitter, and people who are already using generator can migrate without code modification.

I would like that your comments, and @mardy also that require that point.

You think that below like 
//definition
XX_emitter(DATA, const vector<string> clients = {""});
sign::signal<void, DATA, const vector<string>&> XXX_signal

// signal connect
XXX_signal.connect(mem_func(this, XXX_emitter));

// calling
XXX_signal.emit(DATA); // clients will be expected with default param
but you will be show build error required emit(DATA, vector) type 

Default param was decided compile time if it call directly emitter() from user code,
but emitter will be connected by template function's XXX_signal.connect() with decided type
Your comment need to require template default parameter. 
That point must change sigc's connect function of library.

To mardy also wrote like below comments
> Do you mean "emitter(const std::vectorGlib::ustring &destination_bus_names = {}, ...)" default parameter?
> So do you expect "XXX_signal.emit(DATA)" using broadcast and "XXX_signal.emit(ClientList, DATA)" using multicast ?
> 
> But, it will be binding signal to XXX_signal(mem_fun(this, &emitter), then it is required
> clientList parameter again like this "XXX_signal.emit({}, DATA)".
> Because binded XXX_signal.emit() will don't know emitter function's default parameter.
> 
> therefore I patched dedicated signal variable for multicast.

File: codegen_glibmm/templates/stub.h.templ (hunk 6 above)
Comment (liebeym) [reply codegen_glibmm/templates/stub.h.templ]: @mardy 
I was fixed that bug 
Please review again.

**Comment by summeromance** [comment]
@liebeym I have tested destination_bus_names as default parameter as follows:

> //definition
> XX_emitter( DATA, const vector &clients = {} );
> sign::signal<void, DATA, const vector&> XXX_signal
> 
> // call
> XX_emitter(DATA);
> std::vector<Glib::ustring> bus_names;
> bus_names.push_back("aaa");
> bus_names.push_back("bbb");
> XX_emitter(DATA, bus_names);
> 

And there was no build error.

**Comment by liebeym** [comment]
@summeromance 
Thanks your try,  emitter function is not but emit function connected sigc::connect
generated code is used emit function.

emitter function will worked absolutely, it is using  default parameter function.
emit function connected signal is sigc::signal template class’s member function.

**Comment by summeromance** [comment]
@liebeym Ok, now I understand what you are saying.
Let's try this.

```
// define
void XX_emitter(DATA, const std::vector<Glib::ustring> &destination_bus_names);
sigc::signal<void, DATA> XX_signal;
sigc::signal<void, DATA, std::vector<Glib::ustring>&> XX_selectiveSignal;

// connect 
XX_signal.connect(sigc::bind<std::vector<Glib::ustring>>(sigc::mem_fun(this, &Stub::XX_emitter), {""}));
XX_selectiveSignal.connect(sigc::mem_fun(this, &Stub::XX_emitter));

// call
XX_signal(DATA);

std::vector<Glib::ustring> bus_names;
bus_names.push_back("aaa");
bus_names.push_back("bbb");
XX_selectiveSignal(DATA, bus_names);
```

We explicitly use "selectiveSignal" only when we need to emit a signal selectively, otherwise original signal code acts as broadcast. (no changes)

**Comment by liebeym** [comment]
@summeromance 
there is good idea, yes I try like that

**Comment by liebeym** [comment]
@summeromance @mardy 
Please review again
I updated patchset binding emitter for selective signal and passing reference type for emitter

**Comment by summeromance** [comment]
@liebeym  Thank you for accepting my suggestion. 
Regarding the selective signal,  I think it's okay, but now it's time for someone else's opinion.

File: codegen_glibmm/templates/stub.cpp.templ [hunk 7]
```diff
@@ -27,7 +27,9 @@ inline std::string specialGetter(Glib::Variant<std::string> variant)
     m_interfaceName("{{ interface.name }}")
 {
 {% for signal in interface.signals if signal is supported_by_sigc %}
-    {{ signal.name }}_signal.connect(sigc::mem_fun(this, &{{ interface.cpp_class_name_stub }}::{{ signal.name }}_emitter));
+{{ signal.name }}_signal.connect(sigc::bind<0>(sigc::mem_fun(this, &{{ interface.cpp_class_name_stub }}::{{ signal.name }}_emitter),
```
Comment (mardy) [inline codegen_glibmm/templates/stub.cpp.templ:30]: Restore the indentation of this line

File: codegen_glibmm/templates/stub.cpp.templ [hunk 8]
```diff
@@ -191,16 +194,19 @@ void {{ class_name_with_namespace }}::{{ signal.name }}_emitter(
...
+                    "{{ signal.iface_name }}",
+                    "{{ signal.name }}",
+                    bus_name,
+                    params);
+        }
     }
 }
 
 {% endfor %}
+
```
Comment (mardy) [inline codegen_glibmm/templates/stub.cpp.templ:209]: remove this

**Comment by liebeym** [comment]
@mardy 
I updated patch by your comments.
but, CI/CD build has a problem,  there system can't feedback success.

**Comment by liebeym** [comment]
@mardy 
Could you submit this patch?
Very long time spent it :(
